source/on-abort.ts

```diff
@@ -1,47 +1,60 @@
+/* eslint-disable unicorn/no-nested-ternary */
 type Handle =
-	| {disconnect: VoidFunction}
-	| {abort: VoidFunction}
-	| {abortAndReset: VoidFunction}
-	| VoidFunction;
+	| {disconnect(): void}
+	| {abort(): void}
+	| {abortAndReset(): void}
+	| (() => void);
 
-function addListeners(signal: AbortSignal, handles: Handle[]) {
-	// Add one listener per `handle` so that errors don't block other listeners
-	for (const handle of handles) {
-		const options = {once: true};
-		if ('disconnect' in handle) {
-			// Like MutationObserver
-			signal.addEventListener('abort', handle.disconnect.bind(handle), options);
-		} else if ('abort' in handle) {
-			// Like AbortController
-			signal.addEventListener('abort', handle.abort.bind(handle), options);
-		} else if ('abortAndReset' in handle) {
-			// Like ReusableAbortController
-			signal.addEventListener('abort', handle.abortAndReset.bind(handle), options);
-		} else if (typeof handle === 'function') {
-			signal.addEventListener('abort', handle, options);
-		} else {
-			throw new TypeError('Invalid AbortSignal handle type');
-		}
-	}
-}
+const createListener = (handle: Handle): () => void =>
```

Comment:
Slightly-breaking change: function handles would have access to the abort `event` object, even though the `Handle` did not specify this.

I just realized that this had a bad side effect: any abort controllers would be aborted with _an event_ rather than with an error.

On second thought, let me revisit and extract this change into a separate PR